msdm/algorithms/lrtdp.py

```diff
@@ -28,27 +31,47 @@ def __init__(self,
         if max_trial_length is None:
             max_trial_length = float('inf')
         self.max_trial_length = max_trial_length
+        self.event_listener_class = event_listener_class
 
     def plan_on(self, mdp: MarkovDecisionProcess):
+        self._set_up_plan_on()
+        self.lrtdp(
+            mdp, heuristic=self.heuristic, iterations=self.iterations
+        )
+        res = self._tear_down_plan_on(mdp, self.heuristic)
+        return res
+
+    def _set_up_plan_on(self):
         self.res = PlanningResult()
+
         if self.seed is None:
             self.res.seed = random.randint(0, int(2**30))
         else:
             self.res.seed = self.seed
         self.rng = random.Random(self.seed)
-        self.lrtdp(
-            mdp, heuristic=self.heuristic, iterations=self.iterations
-        )
+
+        if self.event_listener_class is not None:
+            self.res.event_listener = self.event_listener_class()
+        else:
+            self.res.event_listener = None
+
+    def _tear_down_plan_on(self, mdp, heuristic):
         res = self.res
+
+        # put together policy, q-values and initial value
         res.policy = {}
         res.Q = defaultdict(lambda : dict())
-        for s in sum(self.res.trials, []) + [state for state, solved in res.solved.items() if solved]:
+        for s, solved in res.solved.items():
```

Comment:
As discussed, to correctly implement anytime behavior, this should include more states. `res.V.keys()` should probably suffice?

Reply:
fixed and tested in e0ce3fd6e1807596dc7c173cfbf2440de1617fa9